### DISCUSSION TITLE
Block Container for Explicit Interface Implementation

### DISCUSSION THREAD

**Discussion by TiberiumOxide** [dotnet/csharplang#5221] [Language Ideas] [answered]
**Problem**
Sometimes classes and structs are forced implement interfaces explicitly due to unforeseen project requirements or project management mistakes. In situations like these I think that the classes become cluttered and hard to read.
**Proposed Solution**
Syntax which groups explicit interface implementation members would make these classes and structs easier to read and reduce verbosity, while making the intent of the design clearer. This could be accomplished with the following syntax:
~~~
class MyClass : InterfaceA
{
    ulong foo;
    string Bar()
    {
         // Do something that returns string
    }

    implement InterfaceA
    {
        ulong Foo { get => foo; set => foo = value; }
        string Bar() => this.Bar();
    }
}
~~~
which is equivalent to:
~~~
class MyClass : InterfaceA
{
    ulong foo;
    string Bar()
    {
         // Do something that returns string
    }
    
    ulong InterfaceA.Foo { get => foo; set => foo = value; }
    string InterfaceA.Bar() => this.Bar();
}
~~~

**Exaggerated Use-Case Scenario**
Suppose a C# developer designed a class called `UsefulResource` which implements the `IUsefulResource` interface. A large code base which depends on the `IUsefulResource` has been developed over the past few weeks and is difficult to modify. The relevant type declarations are shown below:
~~~
interface IUsefulResource
{
    IResourceDescription GetDescription();
    IResourceHandle CheckOut();
}
interface IResourceDescription
{
    ResourceStatusEnum Status { get; }
    ulong NumberOfTimesUsed { get; }
    string UserFriendlyDescription { get; }
}
interface IResourceHandle : IDisposable
{
    //...
    //...
    //...
}

class UsefulResource : IUsefulResource
{
    public IResourceHandle CheckOut() { /*..code..*/  }
    public IResourceDescription GetDescription() { /*..code..*/ }
}
~~~

As a project deadline is only a few days away, the developer is informed that any class which implements `IUsefulResource` must also implement `IAnnoyinglyLongInterfaceFromProjectManager`, which introduces several naming conflicts and necessitates the use of explicit interface implementation members. The `UsefulResource` class now looks like this:

~~~
interface IAnnoyinglyLongInterfaceFromProjectManager
{
    ulong IncidentCounter { get; }
    string Status { get; }
    string GetDescription();
    IDisposable CheckOut();
}
class UsefulResource : IUsefulResource, IAnnoyinglyLongInterfaceFromProjectManager
{
    public IResourceHandle CheckOut() { /*..code..*/  }
    public IResourceDescription GetDescription() { /*..code..*/ }

    // implicit members which only exist to comply with an interface add clutter to the local code base
    public ulong IncidentCounter => GetDescription().NumberOfTimesUsed;
    public string Status => GetDescription().Status.ToString();

    // explicit member declaration is hideous, especially when mixed implicit implementations
    IDisposable IAnnoyinglyLongInterfaceFromProjectManager.CheckOut() => CheckOut();
    string IAnnoyinglyLongInterfaceFromProjectManager.GetDescription() => GetDescription().UserFriendlyDescription;
}
~~~

With the proposed feature, however, the class could look like this instead. All of the changes foisted upon the designer's class can be pushed into a block and segregated from the rest of the class definition. Now it is crystal clear to anyone inspecting the class which members (and possibly conversion code) exist for the sole purpose of complying with an interface.
~~~
class ProposedUsefulResource : IUsefulResource, IAnnoyinglyLongInterfaceFromProjectManager
{
    // Actual code contributed by the class goes here
    public IResourceHandle CheckOut() { /*..code..*/  }
    public IResourceDescription GetDescription() { /*..code..*/ }


    // All code to comply with interface goes here
    implement IAnnoyinglyLongInterfaceFromProjectManager
    {
        ulong IncidentCounter => GetDescription().NumberOfTimesUsed;
        string Status => GetDescription().Status.ToString();
        string GetDescription() => GetDescription().UserFriendlyDescription;
        IDisposable CheckOut() => this.CheckOut();
    }
}
~~~

**Compatibility Considerations**
The proposed feature can be considered "syntactical sugar" because it introduces no new functionality and maps 1-to-1 with existing syntax. Therefore there is no logical conflict with existing language definitions. The "implement" keyword is lowercase and should not conflict with any existing code which uses capital letters to define its types.

**Comment by CyrusNajmabadi** [comment, marked as answer]
I don't see a benefit of this over just using a region it adding another partial part that contains the interface methods.

**Reply by TiberiumOxide** [reply]
Lots of people don't like regions and partial classes. Implementing 3rd party interfaces is a very important part of large designs with multiple programmers and I think the users of C# could benefit from a formal way to dedicate a portion of their class declaration for explicit interface implementation. I think a LOT of development teams would use something like the `implement` block if given the option.

I think that explicitly implementing 3rd party interfaces is such a commonly used design pattern that it deserves some first-class citizen recognition. 

It would also be more convenient to not have to type the name of the interface every single time you want to explicitly implement an interface member.

**Reply by CyrusNajmabadi** [reply]
> Lots of people don't like regions and partial classes.

Inventing a new similar-purpose syntactic system just because someone doesn't like the existing systems isn't really going to be worth it. The existing options are widespread, functional, and multi purpose. A brand new feature to do something less isn't something I think we'd want to do.

**Reply by TiberiumOxide** [reply]
I understand now. I underestimated the burden of proof required to necessitate the introduction of new language features.

**Comment by JustNrik** [comment]
I would personally prefer to have explicit and implicit interface implementation at the same time like VB does. Something like

```cs
public interface IFoo
{
    void DoSomething();
}

public class Foo : IFoo
{
    public void IFoo.DoSomething()
    {
        // ...
    }
}

// would translate into

public class Foo : IFoo
{
    public void DoSomething()
    {
        // ...
    }
    
    void IFoo.DoSomething() => this.DoSomething();
}
```

so we wouldn't even need regions

**Comment by HaloFour** [comment]
Use a partial class, keep those "unforeseen" implementation requirements in a separate location.  I don't believe that C# as a language should add features specifically because you have a disagreement with your project manager.

**Reply by TiberiumOxide** [reply]
The use case given is fictitious and was given as an example to demonstrate why someone might want a formal way to group class members which are not part of the original design. 

I believe that explicitly implementing 3rd party interfaces is an increasingly important part of scaled project development with C#. So much so that it deserves syntactical recognition in the same way C# recognizes events and properties.

**Reply by HaloFour** [reply]
I don't see how anything has changed in the last 20 years to necessitate a reevaluation of the language features in this space.  Explicit implementation has always been important and you have multiple methods through which to group those members if you so choose.

**Reply by CyrusNajmabadi** [reply]
> The use case given is fictitious

Without a good use case, there's no motivation for this.
